source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,183 @@
+
+class CacheInsertChunkBody : public CacheInsertChunk {
+public:
+  CacheInsertChunkBody(const Buffer::Instance& buffer, bool end_stream)
+      : buffer_(buffer), end_stream_(end_stream) {}
+
+  void send(InsertContext& context,
+            std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) override {
+    size_t sz = buffer_.length();
+    context.insertBody(
+        std::move(buffer_),
+        [on_complete, end_stream = end_stream_, sz](bool ready) {
+          on_complete(ready, end_stream, sz);
+        },
+        end_stream_);
+  }
+
+private:
+  Buffer::OwnedImpl buffer_;
+  const bool end_stream_;
+};
+
+class CacheInsertChunkTrailers : public CacheInsertChunk {
+public:
+  explicit CacheInsertChunkTrailers(const Http::ResponseTrailerMap& trailers)
+      : trailers_(Http::ResponseTrailerMapImpl::create()) {
+    Http::ResponseTrailerMapImpl::copyFrom(*trailers_, trailers);
+  }
+
+  void send(InsertContext& context,
+            std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) override {
+    // While zero isn't technically true for the size of trailers, it doesn't
+    // matter at this point because watermarks after the stream is complete
+    // aren't useful.
+    context.insertTrailers(*trailers_, [on_complete](bool ready) { on_complete(ready, true, 0); });
+  }
+
+private:
+  std::unique_ptr<Http::ResponseTrailerMap> trailers_;
+};
+
+CacheInsertQueue::CacheInsertQueue(Http::StreamEncoderFilterCallbacks& encoder_callbacks,
+                                   InsertContextPtr insert_context, AbortInsertCallback abort)
+    : dispatcher_(encoder_callbacks.dispatcher()), insert_context_(std::move(insert_context)),
+      low_watermark_bytes_(encoder_callbacks.encoderBufferLimit() / 2),
+      high_watermark_bytes_(encoder_callbacks.encoderBufferLimit()),
+      encoder_callbacks_(encoder_callbacks), abort_callback_(abort) {}
+
+void CacheInsertQueue::insertHeaders(const Http::ResponseHeaderMap& response_headers,
+                                     const ResponseMetadata& metadata, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  // While zero isn't technically true for the size of headers, headers are
+  // typically excluded from the stream buffer limit.
+  insert_context_->insertHeaders(
+      response_headers, metadata,
+      [this, end_stream](bool ready) { onChunkComplete(ready, end_stream, 0); }, end_stream);
+  chunk_in_flight_ = true;
+}
+
+void CacheInsertQueue::insertBody(const Buffer::Instance& chunk, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  if (chunk_in_flight_) {
+    size_t sz = chunk.length();
+    queue_size_bytes_ += sz;
+    chunks_.push_back(std::make_unique<CacheInsertChunkBody>(chunk, end_stream));
+    if (!watermarked_ && queue_size_bytes_ > high_watermark_bytes_) {
+      if (encoder_callbacks_.has_value()) {
+        encoder_callbacks_.value().get().onEncoderFilterAboveWriteBufferHighWatermark();
```

Comment:
optional: should we stacks track these?

Reply:
sorry, is it worth tracking this as a persistent stat? we track many watermrak events so we can tell where things are backed up.  optional but basically do you think folks will want to monitor this?

Reply:
Filed #28807 to add stats.